### ISSUE TITLE
array attribute defaults: always, not always, checking, how to instantiate

### ISSUE THREAD

**Issue by stscieisenhamer** [STScI-JWST/jwst#1083]
#1075 brought up a couple of issues with datamodels and the handling of the data arrays.  Currently, if an array is specified in the schema, the array will exist upon access, regardless of whether it was part of the initializing data.  From #1075, some issues with this approach have arisen:

1. Allow "optional" arrays. These arrays would not automatically instantiate when accessed. If an array was not in the initializing data, or explicitly set, an `AttributeError` should be thrown. This behavior could key off the existence of attribute's `default` property, or create another property.
2. Ability to check if an array exists without instantiating. This can currently be done using the `DataModel.instance`, but that is rather hacky/api-breaking.
3. Specify how an array is instantiated on a case-by-case basis. If an attribute is specified with `ndim` less than the primary data, only the _fastest moving_ dimensions are used in instantiating the array. However, as an example, `Level1bModel.zeroframe` does not follow this model, instead needing an arbitrary subset of the main data's dimensions. Note that this may be the only case and not worth consideration. It is mentioned here in case there are others, and that the current solution may break if the first two issues are dealt with.

Issue note: This is +1 on the esoteric scale. Just documenting here for later consideration.

**Comment by hbushouse** [comment]
This could potentially be expanded to also include meta attributes, not just arrays. Right now all meta attributes are effectively optional: if a FITS file that's being loaded doesn't contain the defined keyword, the attribute is left set to None. This also means that the keyword will not appear in any output file created from the data model. Which is fine for keywords that aren't required. But it would be useful to be able to indicate whether a given keyword/meta attribute is required and hence throw an error (or at least a warning) when it doesn't appear in a file that's being loaded.

**Comment by bernie-simon** [comment]
A big reason why datamodels works the way it currently does is that when reading and writing a model it walks the data and not the schema. This means that a new field in the schema such as "optional" or "required" is not going to be handled if the data item corresponding to the schema field is not already there. And although handling reading and writing by going through the schema validator is clever and elegant in one way, it's a pain in the neck in another. It makes debugging problems in the code difficult, as reading and writing through callbacks triggered by the schema fields makes understanding the path of code execution difficult. When the code is obscure, bugs
are not caught. So while adding functionality to the schema may seem attractive, for these reasons I believe it would be better to handle the requested changes through Python code and not by enhancing the schema.

Optional arrays: This can be handled with the current code by first checking for the existence of the array with hasattr:

    if hasattr(model, "foobar"):
        array = model.foobar
    else:
        raise AttributeError("Foobar not found")

I could modify the code in properties.py to raise an AttributeError if there is no default defined for an array, but I hesitate to do this, as it's an interface change and may break existing code and I don't
know a real use case for it. But if people feel it's a needed change, it's an easy change.

Check if an array exists: You can do this with hasattr(), as in the previous example code.

Handle array instantiation on a case by case basis: This is best handled by checking for the existence of an array and having specific code to handle the instantiation if it's not found.

Throw a warning or error if a keyword is not found when reading: Keyword checking can be added to the __init__ method of any class that requires it. If the need is widespead, a check_required() method can be added to model_base.py. The method would take a tuple of strings as an argument and check for the existence of each in the model metadata. It would be called by the __init__ method. The code is simple if it only needs to check the metadata:

    def check_required(self, fields):
        missing = []
	for field in fields:
	    if field not in self.meta:
	        missing.append(field)
	if len(missing) > 0:
	    raise AttributeError("Missing keywords: " + str(missing))

**Comment by bernie-simon** [comment]
Ok, I take back what I said about marking parameters as required in the schema. Because of the way the code is written, it's possible to throw an error if a fits keyword is missing. I'll be adding code to support the keyword fits_required, which will take a value, either true or false.

**Comment by hbushouse** [comment]
Is it possible to extend such logic to schema attributes that are data arrays (not just keywords)? By default everything seems to be considered optional, in that if a defined array/extension is not found in a FITS file it's simply left set to None in the resulting data model with no complaints or warnings.

**Comment by bernie-simon** [comment]
Yes, the same code can check for both data and keywords missing from the schema.

**Comment by bernie-simon** [comment]
I've reviewed the comments on this issue and #1075. After reviewing, I'd like to make two changes to how datamodels works currently, but would like your opinion.

Add a hasattr method to ObjectNode, which would simply be an alias for _ _hasattr_ _. The Python hasattr function will create the value, as it calls _ _getattr_ _, which creates values from defaults. The hasattr method does not, but it will say an attribute exists if there is a default for it in the schema. To test if an attribute exists, regardless of whether there is a default in the schema, use  a test like 

    attr in im.meta.instance

Only create a value when the schema contains a default field. Currently datamodels will always create a default for an array, but only create a default for a scalar if there is a default. I don't understand the reason for this distinction and would like to remove it.

**Comment by hbushouse** [comment]
I guess the fact that datamodels currently always creates a default for an array is what allows the weird statements like:
`self.dq = self.dq`
in model modules like RampModel and ImageModel to work, such that if there wasn't a dq array in the file being loaded, the effect of that statement is to cause one to be created in the model and have it populated with the default value?

If we were to remove the behavior that creates a default array even when a default is not specified, then we would just have to make sure we actually do have a default specified, when appropriate, in each of the data models (e.g. for RampModel where dq and err usually aren't in the input file).

Then the way to force either an array or scalar (keyword) to be required is to simply not specify a default for them. If there's no default specified, then that array or scalar won't appear in the model if they aren't contained in an input file. But if we really want an array or scalar to be required, the action should do more than just quietly not create an entry for it in the model. It should throw at least a warning, if not an error, because we've said that item must be there. It should only move on quietly for cases where we've decided the item is not required.

**Comment by philhodge** [comment]
It would be great to get rid of statements such as `self.dq = self.dq`; they're really obscure.

If we're opening an input file with `model = datamodels.open(filename)`, how could we specify a default for, say, `dq`?  We won't (usually) know the shape or data type needed before opening the file, and we might not even know whether the file will or should contain `dq` as opposed to `pixeldq` or `groupdq`, for example.  If we're creating a model from scratch, that's another story.

If we open an input file, I think we should get (in the returned model) only what's in the input file, except for attributes with default values specified in the schema (e.g. keywords).  After opening the file, we can check what type the model is, and then we'll know what attributes we expect and what we need.  There could be methods of the model to explicitly create default arrays and assign them to attributes.  Those methods would be no-ops if the attributes already exist.  Thus if a step finds that it has opened a file and has gotten a RampModel, for example, the step then knows that it will need a `groupdq` attribute, so the step could just call the method to create an appropriate default for that attribute, and the step would never have to explicitly check whether that attribute exists.

**Comment by bernie-simon** [comment]
The purpose of lines like

    self.dq = self.dq

in the datamodels classes is to initialize the array to its default if it is not already defined. The lines are admittedly obscure, a comment could be added to explain them. 

The shape of the dq or other array initialize to the default is the same as the shape of the data array unless ndim is defined in its schema. In that case the shape is the last ndim dimensions of the shape. 

The purpose of initialzing arrays to defaults when they are not defined otherwise is to save downstream code from needing to test for their existence. I think the current code handles initializing arrays from defaults well. (No ego here, Mike wrote the code, not me.) I think the __init__ method is the correct place to handle default initialization. If there are cases that the current code does not handle well, the code can be modified or rewritten.  I'm not clear why the current code can't handle pixeldq or groupdq as well as dq. The schema will have entries for groupdq and pixeldq and the __init__ method will initialize them. Look at datamodels/ramp.py for an example of how this is done.

**Comment by philhodge** [comment]
I agree that the initialization works well for the case that you really want all those arrays, but some of the time a good fraction of the arrays are never touched.  I was just suggesting a way to avoid initializing arrays unless the step actually needs them.

I also think the old HST convention for FITS image extensions with constant values (i.e. just writing the header) was a good idea, and we could use that convention for JWST data.
